base/threadingconstructs.jl

```diff
@@ -18,65 +18,27 @@ on [`threadid()`](@ref).
 """
 nthreads() = Int(unsafe_load(cglobal(:jl_n_threads, Cint)))
 
-function _threadsfor(iter,lbody)
-    lidx = iter.args[1]         # index
-    range = iter.args[2]
-    quote
-        local threadsfor_fun
-        let range = $(esc(range))
-        function threadsfor_fun(onethread=false)
-            r = range # Load into local variable
-            lenr = length(r)
-            # divide loop iterations among threads
-            if onethread
-                tid = 1
-                len, rem = lenr, 0
-            else
-                tid = threadid()
-                len, rem = divrem(lenr, nthreads())
-            end
-            # not enough iterations for all the threads?
-            if len == 0
-                if tid > rem
-                    return
-                end
-                len, rem = 1, 0
-            end
-            # compute this thread's iterations
-            f = firstindex(r) + ((tid-1) * len)
-            l = f + len - 1
-            # distribute remaining iterations evenly
-            if rem > 0
-                if tid <= rem
-                    f = f + (tid-1)
-                    l = l + tid
-                else
-                    f = f + rem
-                    l = l + rem
+function _threadsfor(iter_stmt, lbody)
+    loopvar   = iter_stmt.args[1]
+    iter      = iter_stmt.args[2]
+    rng = gensym(:rng)
+    out = quote
+        @sync for $rng in $(Iterators.partition)($iter, $(length)($iter) ÷ $(nthreads)())
+            Base.Threads.@spawn begin
+                for $loopvar in $rng
```

Comment:
Strictly speaking, I think this is a breaking change as the following code will behave differently before and after this PR:

```julia
@threads for ...
    @spawn ...
end
```

Before this PR, `@spawn` is not synchronized at the end of `@threads` block. After this PR, it is not thread-safe: https://github.com/JuliaLang/julia/issues/34666. You can make it thread-safe by adding `@sync` before `for $loopvar in $rng` although the behavior of `@spawn` is still different (it's synchronized by `@threads`).

Reply:
Mhm, this could conceivably  change behaviour (though threading is experimental, so breaking changes are allowed), but before if you did
```julia
@threads for ...
    @spawn ...
end
```
then the `@spawn` would destructively interfere with the tasks spawned by `@threads`.


Are you able to cook up a valid test case that would have passed on the old version of `@threads` but would fail on this version?

Reply:
Ah I see an error in the test suite provides an example of exactly what you referenced.

Reply:
> (though threading is experimental, so breaking changes are allowed)

Yeah, I think I agree. Though it could be surprising if `@threads` acts implicitly as `@sync`.

FWIW, I think `Threads.foreach` would be a much better API as it doesn't hide the fact that you'll have a closure. It's also easy to add keyword options, e.g., for controlling the base case size for better load balancing. Something like this:

```julia
function Threads.foreach(f, xs::AbstractArray; basesize = length(xs) ÷ Threads.nthreads())
    @sync for p in Iterators.partition(xs, basesize)
        @spawn foreach(f, p)
    end
    return
end
```

(I'm experimenting this API here: https://github.com/tkf/ThreadsX.jl/blob/1b2f2a42094986361785b617c064e5fe8cda93af/src/map.jl#L26)

`@threads` can exist as a simple wrapper around `Threads.foreach` for backward compatibility.  This trivially avoids the `@sync` scoping problem.